### PR TITLE
LSP: Handle rpc RequestCancelled specifically.

### LINKED ISSUE #11515: LSP client shows error messages for cancellation request acknowledgements
Version: NVIM v0.5.0-226-gb1e4ec1c2

Based on a conversation in the gitter channel with @norcalli I believe that there is an oversight with the LSP client in its current state.

I was attempting to get `javascript-typescript-langserver` working with LSP, and while I managed to get it working, I would occasionally get strange error messages mentioning `NO_RESULT_CALLBACK_FOUND`.  After some digging, I found out that the language server appears to respond to cancellation requests with a cancellation acknowledgement error code as shown [here](https://github.com/sourcegraph/javascript-typescript-langserver/blob/d107424f1f964c4cb3b6061570890639fd69b9e6/src/connection.ts#L211).  This behavior is apparently [endorsed by the LSP standard](https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/#cancelRequest) with a specific error code set aside for it, though the standard stops short of mandating its use, instead also allowing servers to return partial results.  This error code is even [already defined](https://github.com/neovim/neovim/blob/d5aaad14ecdd2047089e1a018e97af1f790b3e42/runtime/lua/vim/lsp/protocol.lua#L164) in the protocol file that neovim uses for LSP, though it's not actually used anywhere.

It seems to me like if neovim sends a cancellation request, it probably shouldn't show a verbose error message that interrupts the user when the acknowledgement "error" comes back - that seems like a bug to me.  After a conversation with @norcalli it was suggested that perhaps a terse one-line message could be displayed instead that doesn't interrupt the user that mentions that the request timed out and had to be cancelle.  Or perhaps nothing at all.  There were also suggestions of being able to override the timeout of the request, both globally and per-client.

### PR DESCRIPTION
**Pull request by norcalli**
This was creating extra noise in errors that we should've been handling
internally.

Fixes #11515

### CODE REVIEW

File: runtime/lua/vim/lsp/rpc.lua [hunk 1]
```diff
@@ -377,6 +378,23 @@ local function create_and_start_client(cmd, cmd_args, handlers, extra_spawn_para
       decoded.error = convert_NIL(decoded.error)
       decoded.result = convert_NIL(decoded.result)
 
+      -- For RequestCancelled, we want to ignore the handlers since this is an
+      -- RPC internal handling code and we shouldn't report this to users.
```
Comment (justinmk) [inline runtime/lua/vim/lsp/rpc.lua]: nit (but also in general): less-verbose wording:

    -- Do not surface RequestCancelled to users, it is RPC-internal.

**Review by justinmk — APPROVED** [review]

File: runtime/lua/vim/lsp/rpc.lua [hunk 2]
```diff
@@ -1,3 +1,4 @@
+local vim = vim
```
Comment (h-michael) [inline runtime/lua/vim/lsp/rpc.lua]: Why did you need to add this line?

Comment (norcalli) [reply runtime/lua/vim/lsp/rpc.lua]: It brings vim into the local scope instead of the global scope which is more efficient. Looking up a global value means hitting the global table, whereas a local value is an upvalue captured by the functions. It also hides a lot of the errors I get from my Lua LSP.

Comment (justinmk) [reply runtime/lua/vim/lsp/rpc.lua]: It's a micro-optimization.
